File: neo/backend/src/app/schemas/agent.py
"""Agent Pydantic schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


class AgentSkill(BaseModel):
    """Agent skill reference."""

    id: int = Field(..., description="Skill ID")
    version: str | None = Field(None, description="Skill version")


class AgentConfig(BaseModel):
    """Agent runtime configuration."""

    temperature: float = Field(default=0.7, description="Model temperature")
    max_tokens: int = Field(default=4096, description="Max output tokens")
    thinking: str = Field(default="low", description="Thinking depth: low/medium/high")
    timeout: int = Field(default=60, description="Timeout in seconds")
    retry: dict = Field(default_factory=dict, description="Retry configuration")


# ============ Request Schemas ============


class AgentCreate(BaseModel):
    """Schema for creating a new Agent."""

    name: str = Field(..., min_length=1, max_length=32, description="Agent name (unique within workspace)")
    description: str | None = Field(None, max_length=500, description="Description")
    prototype_id: int = Field(..., description="Prototype ID")
    prototype_version: str = Field(..., description="Prototype version")
    model: str | None = Field(None, max_length=64, description="Model (inherits from prototype if not specified)")
    skills: list = Field(default_factory=list, description="Enabled skills (list of strings or dicts)")
    config: dict = Field(default_factory=dict, description="Runtime configuration")


class AgentUpdate(BaseModel):
    """Schema for updating an Agent."""

    name: str | None = Field(None, min_length=1, max_length=32, description="Agent name")
    description: str | None = Field(None, max_length=500, description="Description")
    model: str | None = Field(None, max_length=64, description="Model")
    skills: list | None = Field(None, description="Enabled skills")
    config: dict | None = Field(None, description="Runtime configuration")


class AgentListQuery(BaseModel):
    """Schema for listing Agents with filters."""

    page: int = Field(default=1, ge=1, description="Page number")
    page_size: int = Field(default=20, ge=1, le=100, description="Items per page")
    status: str | None = Field(None, description="Filter by status: enabled/disabled")
    prototype_id: int | None = Field(None, description="Filter by prototype")
    search: str | None = Field(None, max_length=100, description="Search in name/description")


# ============ Response Schemas ============


class AgentResponse(BaseModel):
    """Schema for Agent response."""

    id: int
    name: str
    description: str | None
    prototype_id: int
    prototype_version: str
    workspace_id: int
    model: str
    skills: list  # Can be list of dicts or list of strings
    config: dict
    status: str
    created_by: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class AgentListResponse(BaseModel):
    """Schema for Agent list response with pagination."""

    items: list[AgentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class AgentStatusResponse(BaseModel):
    """Schema for Agent status update response."""

    id: int
    status: str
    updated_at: datetime

    model_config = {"from_attributes": True}

File: neo/frontend/lib/api/agent.ts
/**
 * Agent API client
 * Connects to backend API via frontend proxy: /api/v1/workspaces/{workspace_code}/agents
 */

import { getAuthHeaders } from "@/lib/utils/auth";

// API base URL - use relative path (Next.js API route will proxy)
const API_BASE = "";

// Types matching backend API
export type AgentStatus = "enabled" | "disabled" | "deleted";

/**
 * Reference to an enabled skill on an Agent.
 *
 * 与后端 AgentSkill 模型对齐：使用 `skill_id`（前端约定名），后端在
 * AgentUpdate / AgentCreate 时以 `id` 字段存取（见 backend/src/app/services/agent_service.py）。
 */
export interface AgentSkillRef {
  skill_id: number;
  version: string;
}

export interface AgentPrototypeInfo {
  id: number;
  code: string;
  name: string;
  version: string;
}

export interface AgentResponse {
  id: number;
  name: string;
  description: string | null;
  prototype_id: number;
  prototype_version: string;
  workspace_id: number;
  model: string;
  skills: AgentSkillRef[];
  config: Record<string, unknown>;
  status: AgentStatus;
  created_by: number;
  created_at: string;
  updated_at: string;
  prototype: AgentPrototypeInfo | null;
}

export interface AgentListResponse {
  items: AgentResponse[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CreateAgentRequest {
  name: string;
  description?: string;
  prototype_id: number;
  prototype_version: string;
  model?: string;
  skills?: AgentSkillRef[];
  config?: Record<string, unknown>;
}

export interface UpdateAgentRequest {
  name?: string;
  description?: string;
  model?: string;
  skills?: AgentSkillRef[];
  config?: Record<string, unknown>;
  prototype_version?: string;
}

export interface AgentStatusResponse {
  id: number;
  status: AgentStatus;
  updated_at: string;
}

// Error handling
export class ApiError extends Error {
  constructor(
    message: string,
    public code: number,
    public statusCode: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

// Helper function for API requests
async function apiRequest<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<T> {
  const url = `${API_BASE}${endpoint}`;
  const authHeaders = getAuthHeaders();

  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...authHeaders,
      ...options.headers,
    },
    credentials: "include",
  });

  const data = await response.json();

  if (!response.ok) {
    throw new ApiError(
      data.message || "Request failed",
      data.code || response.status,
      response.status,
    );
  }

  if (data.code !== 0) {
    throw new ApiError(data.message, data.code, 200);
  }

  return data.data as T;
}

// ==================== Agent API ====================

/**
 * List all agents in a workspace with pagination and filters
 */
export async function listAgents(
  workspaceCode: string,
  params?: {
    page?: number;
    page_size?: number;
    status?: string;
    prototype_id?: number;
    search?: string;
  },
): Promise<AgentListResponse> {
  const searchParams = new URLSearchParams();

  if (params?.page) searchParams.set("page", params.page.toString());
  if (params?.page_size)
    searchParams.set("page_size", params.page_size.toString());
  if (params?.status) searchParams.set("status", params.status);
  if (params?.prototype_id)
    searchParams.set("prototype_id", params.prototype_id.toString());
  if (params?.search) searchParams.set("search", params.search);

  const queryString = searchParams.toString();
  const endpoint = `/api/v1/workspaces/${workspaceCode}/agents${
    queryString ? `?${queryString}` : ""
  }`;

  return apiRequest<AgentListResponse>(endpoint);
}

/**
 * Get a single agent by ID
 */
export async function getAgent(
  workspaceCode: string,
  agentId: number,
): Promise<AgentResponse> {
  return apiRequest<AgentResponse>(
    `/api/v1/workspaces/${workspaceCode}/agents/${agentId}`,
  );
}

/**
 * Create a new agent
 */
export async function createAgent(
  workspaceCode: string,
  data: CreateAgentRequest,
): Promise<AgentResponse> {
  return apiRequest<AgentResponse>(
    `/api/v1/workspaces/${workspaceCode}/agents`,
    {
      method: "POST",
      body: JSON.stringify(data),
    },
  );
}

/**
 * Update an agent
 */
export async function updateAgent(
  workspaceCode: string,
  agentId: number,
  data: UpdateAgentRequest,
): Promise<AgentResponse> {
  return apiRequest<AgentResponse>(
    `/api/v1/workspaces/${workspaceCode}/agents/${agentId}`,
    {
      method: "PUT",
      body: JSON.stringify(data),
    },
  );
}

/**
 * Delete an agent (soft delete)
 */
export async function deleteAgent(
  workspaceCode: string,
  agentId: number,
): Promise<{ message: string }> {
  return apiRequest<{ message: string }>(
    `/api/v1/workspaces/${workspaceCode}/agents/${agentId}`,
    {
      method: "DELETE",
    },
  );
}

/**
 * Enable an agent
 */
export async function enableAgent(
  workspaceCode: string,
  agentId: number,
): Promise<AgentStatusResponse> {
  return apiRequest<AgentStatusResponse>(
    `/api/v1/workspaces/${workspaceCode}/agents/${agentId}/enable`,
    {
      method: "PATCH",
    },
  );
}

/**
 * Disable an agent
 */
export async function disableAgent(
  workspaceCode: string,
  agentId: number,
): Promise<AgentStatusResponse> {
  return apiRequest<AgentStatusResponse>(
    `/api/v1/workspaces/${workspaceCode}/agents/${agentId}/disable`,
    {
      method: "PATCH",
    },
  );
}
